csharp/ql/src/semmle/code/csharp/dataflow/internal/FrameworkDataFlowImpl.qll

```diff
@@ -0,0 +1,171 @@
+    )
+    or
+    this = TNilContentList() and
+    result = "<empty>"
+  }
+}
+
+/** Provides predicates for constructing content lists. */
+module ContentList {
+  /** Gets the empty content list. */
+  ContentList empty() { result = TNilContentList() }
+
+  /** Gets a singleton content list containing `c`. */
+  ContentList singleton(Content c) { result = TConsContentList(c, TNilContentList()) }
+
+  /** Gets the content list obtained by consing `head` onto `tail`. */
+  ContentList cons(Content head, ContentList tail) { result = TConsContentList(head, tail) }
+}
+
+private newtype TFlowSource = MkFlowSource(FlowSource::Range r)
+
+/** A flow-source specification. */
+class FlowSource extends MkFlowSource {
+  private FlowSource::Range r;
+
+  FlowSource() { this = MkFlowSource(r) }
+
+  predicate isSubsetOf(FlowSource fs) { r.isSubsetOf(fs) }
+
+  string toString() { result = r }
+}
+
+module FlowSource {
+  abstract class Range extends string {
+    bindingset[this]
+    Range() { any() }
+
+    predicate isSubsetOf(FlowSource fs) { none() }
+
+    FlowSource toFlowSource() { result = MkFlowSource(this) }
+  }
+}
+
+private newtype TFlowSink = MkFlowSink(FlowSink::Range r)
+
+/** A flow-sink specification. */
+class FlowSink extends MkFlowSink {
+  private FlowSink::Range r;
+
+  FlowSink() { this = MkFlowSink(r) }
+
+  predicate isSubsetOf(FlowSink fs) { r.isSubsetOf(fs) }
+
+  string toString() { result = r }
+}
+
+module FlowSink {
+  abstract class Range extends string {
+    bindingset[this]
+    Range() { any() }
+
+    predicate isSubsetOf(FlowSink fs) { none() }
+
+    FlowSink toFlowSink() { result = MkFlowSink(this) }
+  }
+}
+
+/** A data-flow model for a given framework. */
+abstract class FrameworkDataFlow extends string {
+  bindingset[this]
+  FrameworkDataFlow() { any() }
+
+  /**
+   * Holds if data may flow from `input` to `output` when calling `c`.
+   *
+   * `inputContents` describes the contents that is popped from the access
+   * path from the input and `outputContents` describes the contents that
+   * is pushed onto the resulting access path.
+   *
+   * `preservesValue` indicates whether this is a value-preserving step
+   * or a taint-step.
+   */
+  pragma[nomagic]
+  predicate hasSummary(
+    FrameworkCallable c, SummaryInput input, ContentList inputContents, SummaryOutput output,
+    ContentList outputContents, boolean preservesValue
+  ) {
+    none()
+  }
+
+  /**
+   * Holds if the content list obtained by consing `head` onto `tail` is needed
+   * for a summary specified by `hasSummary()`.
+   *
+   * This predicate is needed for QL technical reasons only (the IPA type used
+   * to represent content lists needs to be bounded).
+   *
+   * Only summaries using content lists of length >= 2 need to override this
+   * predicate.
+   */
+  pragma[nomagic]
+  predicate requiresContentList(Content head, ContentList tail) { none() }
```

Comment:
Do we need content lists of length >= 2 from day one? If not, I suggest we remove that feature from the interface. When we come to need it, the requirements may be different than what we think they are now.

Reply:
We use it for C# already.